src/dataframe/dataframe.jl

```diff
@@ -1344,16 +1344,22 @@ function Base.push!(df::DataFrame, row::Union{AbstractDict, NamedTuple};
     end
 
     old_row_type = typeof(row)
-    if row isa AbstractDict && all(x -> x isa AbstractString, keys(row))
-        row = (;(Symbol.(keys(row)) .=> values(row))...)
+    if row isa AbstractDict
+        if !(keytype(row) <: Symbol)
+            if keytype(row) <: AbstractString || all(x -> x isa AbstractString, keys(row))
+                row = (;(Symbol.(keys(row)) .=> values(row))...)
+            end
+        end
```

Comment:
Why not a single `if`?
```suggestion
    if row isa AbstractDict && keytype(row) !== Symbol &&
        (keytype(row) <: AbstractString || all(x -> x isa AbstractString, keys(row)))
        row = (;(Symbol.(keys(row)) .=> values(row))...)
```